src/auth/components/oidcCallback/OidcCallback.tsx

```diff
@@ -14,47 +19,73 @@ function OidcCallback({
 }: RouteChildrenProps): React.ReactElement | null {
   const { t } = useTranslation();
   const redirectToErrorPage = useErrorPageRedirect();
+  const generigErrorString = 'authentication.genericError.message';
 
-  useEffect(() => {
-    authService
-      .endLogin()
-      .then(() => {
+  const onSuccess = () => {
+    // Handle successful login - redirect to profile page
+    function delay(time: number, callback: () => void) {
+      setTimeout(callback, time);
+    }
+
+    let apiTokens = null;
+
+    // When onSuccess is called getApiTokensFromStorage is not yet available and profile cannot be fetched
+    function checkApiTokens() {
+      apiTokens = getApiTokensFromStorage(); // Call your function and wait for its result
+
+      if (apiTokens === null) {
+        delay(10, checkApiTokens);
```

Comment:
If something goes wrong, I'm guessing there is a possiblity that tokens are never available? 🤔 Maybe there should be an option to end the loop and throw error after x amount of tries 🤔

Reply:
Matti used a pretty nice solution in PSA 👀  https://github.com/City-of-Helsinki/pysakoinnin-sahk-asiointi-ui/pull/169/files#diff-3dabf155febbf2775b0b04a4b659514541a8cba8e2674c4d2b0a8b3bc41f8966R17

Reply:
Moved the whole thing from callback to the client.ts. It's very similar now.